lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (C) 2016 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package fswatcher
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+	"github.com/zillode/notify"
+
+	"github.com/syncthing/syncthing/lib/events"
+)
+
+type FsEvent struct {
+	path string
+}
+
+type FsWatcher struct {
+	folderPath string
+	notifyModelChan chan<- map[string]FsEvent
+	fsEvents map[string]FsEvent
+	fsEventChan <-chan notify.EventInfo
+	WatchingFs bool
+	notifyDelay time.Duration
+	notifyTimer time.Timer
+	notifyTimerNeedsReset bool
+}
+
+func NewFsWatcher(folderPath string) *FsWatcher {
+	return &FsWatcher{
+		folderPath: folderPath,
+		notifyModelChan: nil,
+		fsEvents: make(map[string]FsEvent),
+		fsEventChan: nil,
+		WatchingFs: false,
+		notifyDelay: fastNotifyDelay,
+		notifyTimerNeedsReset: false,
+	}
+}
+
+func (watcher *FsWatcher) StartWatchingFilesystem() (<-chan map[string]FsEvent, error) {
+	notifyModelChan := make(chan map[string]FsEvent)
+	fsEventChan, err := setupNotifications(watcher.folderPath)
+	if fsEventChan != nil {
+		watcher.WatchingFs = true
+		watcher.fsEventChan = fsEventChan
+		go watcher.watchFilesystem()
+	}
+	watcher.notifyModelChan = notifyModelChan
+	return notifyModelChan, err
+}
+
+func ExtractChangedPaths(events map[string]FsEvent) []string {
+	var paths []string
+	for path := range events {
+		paths = append(paths, path)
+	}
+	return paths
+}
+
+var maxFiles = 512
+
+func setupNotifications(path string) (chan notify.EventInfo, error) {
+	c := make(chan notify.EventInfo, maxFiles)
+	if err := notify.Watch(path, c, notify.All); err != nil {
```

Comment:
What I can do is call [notify.Close()](https://github.com/Zillode/notify/blob/79dcc1b89960622149f3b106bcc4e3804a2c8b27/notify.go#L67) on it, although that doesn't seem to close the channel either.